## DAB200 -- Lab 2 - Group 10



### Part 0

Please provide the following information:
 - Group Number  - 10
 - Group Members
     1. Keshav Yadav 
     2. Sai Sankeerth Koduru
     3. Gurpinder Kaur
 
     

### Part 1 - Create and evaluate an initial model

#### Code (15 marks)

In [77]:
#Code added by Keshav[10-03-2021] - Importing the pandas package
import pandas as pd


In [78]:
#Code added by Keshav[10-03-2021] - Importing our file to create a dataframe house_rent
house_rent = pd.read_csv('rent_2.csv')
#Code added by Keshav[10-03-2021] - Displaying the transpose of the house_rent head to check the number of columns
house_rent.head(5).T

0  \
Unnamed: 0                                                       0   
bathrooms                                                      1.0   
bedrooms                                                         2   
building_id                                                      0   
created                                        2016-05-07 01:32:32   
description      NEW TO MARKET! XL Gorgeously Gut Renovated Two...   
display_address                                   East 11th Street   
features               ['Pre-War', 'Dogs Allowed', 'Cats Allowed']   
latitude                                                   40.7269   
longitude                                                 -73.9787   
manager_id                        35dc123b6f5f9ea6d7b755f21b9c3585   
photos           ['https://photos.renthop.com/2/6977294_198d4cf...   
price                                                         2400   
street_address                                620 East 11th Street   
interest_level                                                 low   
mgr_apt_count                                                   21   

                                                                 1  \
Unnamed: 0                                                       1   
bathrooms                                                      1.0   
bedrooms                                                         0   
building_id                       cbbc47cf64b2ee7c5504589d7b4cf68b   
created                                        2016-05-12 06:16:18   
description      This Bright And Spacious 2BED Apartment Featur...   
display_address                                        W 51 Street   
features         ['Roof Deck', 'Elevator', 'Garden/Patio', 'Lau...   
latitude                                                   40.7644   
longitude                                                 -73.9891   
manager_id                        2637547481b1d0b116843dcd4cf94211   
photos           ['https://photos.renthop.com/2/7002794_1c1acfa...   
price                                                         2500   
street_address                                     407 W 51 Street   
interest_level                                                 low   
mgr_apt_count                                                   24   

                                                                 2  \
Unnamed: 0                                                       2   
bathrooms                                                      2.0   
bedrooms                                                         3   
building_id                       03c36303e8adcc78b395f6c02cdff497   
created                                        2016-04-21 01:01:41   
description      Come view this nice 3 bedroom in Prime Bushwic...   
display_address                                    Bushwick Avenue   
features         ['gym', 'Lounge', 'Laundry In Building', 'Virt...   
latitude                                                   40.6933   
longitude                                                 -73.9255   
manager_id                        66f8f5a9fff474db421459fd2bfb3032   
photos           ['https://photos.renthop.com/2/6903256_633f000...   
price                                                         3184   
street_address                                 894 Bushwick Avenue   
interest_level                                                 low   
mgr_apt_count                                                    7   

                                                                 3  \
Unnamed: 0                                                       3   
bathrooms                                                      1.0   
bedrooms                                                         3   
building_id                       6b3fbd289f786a2fdac1bf8aed04060e   
created                                        2016-06-22 04:31:24   
description      Amazing 3 bedroom apartment in Midtown East. P...   
di

In [79]:
#Code added by Keshav[10-03-2021] - Check the information about the data frame and like column name, data type and non-null count
house_rent.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       20000 non-null  int64  
 1   bathrooms        20000 non-null  float64
 2   bedrooms         20000 non-null  int64  
 3   building_id      20000 non-null  object 
 4   created          20000 non-null  object 
 5   description      19423 non-null  object 
 6   display_address  19947 non-null  object 
 7   features         20000 non-null  object 
 8   latitude         20000 non-null  float64
 9   longitude        20000 non-null  float64
 10  manager_id       20000 non-null  object 
 11  photos           20000 non-null  object 
 12  price            20000 non-null  int64  
 13  street_address   19994 non-null  object 
 14  interest_level   20000 non-null  object 
 15  mgr_apt_count    20000 non-null  int64  
dtypes: float64(3), int64(4), object(9)
memory usage: 2.4+ MB


In [80]:
#Code added by Keshav[10-03-2021] - Creating the new dataframe of having only numeric feature
house_rent_num = house_rent[['bathrooms', 'bedrooms', 'longitude', 'latitude', 'price']]

house_rent_num.head(3)

bathrooms  bedrooms  longitude  latitude  price
0        1.0         2   -73.9787   40.7269   2400
1        1.0         0   -73.9891   40.7644   2500
2        2.0         3   -73.9255   40.6933   3184

In [81]:
#Code added by Keshav[10-03-2021] - Checking that is there any null values in the numeric dataframe
house_rent_num.isnull().any()

bathrooms    False
bedrooms     False
longitude    False
latitude     False
price        False
dtype: bool

In [82]:
#Code added by Keshav[10-03-2021] - Creating feature vector and Target vector
X = house_rent_num.drop('price', axis=1)
y = house_rent_num['price']

In [83]:
#Code added by Keshav[10-03-2021] - Importing all the necessary packages
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score
import numpy as np
import random

In [84]:
#Code added by Keshav[10-03-2021] - Spliting our datasets into training dataset and testing[validation] datasets
X_Train, X_Test, y_Train, y_Test = train_test_split(X, y, test_size=0.2)  
#Code added by Keshav[10-03-2021] - Creating a random forest regressor
rf = RandomForestRegressor(n_estimators=100, n_jobs=-1, oob_score=True) 
#Code added by Keshav[10-03-2021] - Fitting our model using training dataset 
rf.fit(X_Train, y_Train) 

rf.score(X_Train, y_Train)

0.8560887268298699

In [85]:
#Code added by Keshav[10-03-2021] - calculating the out-of-bag score
rf.oob_score_

0.0038250463664277756

In [86]:
rf.score(X_Test, y_Test)

0.8064587156174415

In [87]:
#Code added by Keshav[10-03-2021] - Calculating the average price for the training uncleaed data and R-Squared score
avg_preds = [np.mean(y_Train)] * len(y_Test)

print(np.mean(y_Train))

r2_score(y_Test, avg_preds) 

3945.2443125


-0.036297147215820225

In [91]:
#Code added by Keshav[10-03-2021] - Creating Empty lists for storing the values for testing and training fields respectivelly
Train_r2 = []
Train_mae = []
Test_r2 = []
Test_mae = []
oob_scores = []
#Code added by Keshav[10-03-2021] - Creating a for loop to calculate the values of R-Squared, MAE for test and train and Out-of-bag score for 10 runs for uncleaned data
for i in range(10):
    X_Train, X_Test, y_Train, y_Test = train_test_split(X, y, test_size=0.2)
    rf = RandomForestRegressor(n_estimators=100, n_jobs=-1, oob_score=True) 
    rf.fit(X_Train, y_Train)
    train_preds = rf.predict(X_Train)
    val_preds = rf.predict(X_Test)
    Train_r2.append(round(r2_score(y_Train, train_preds), 2))
    Test_r2.append(round(r2_score(y_Test, val_preds), 2))
    Train_mae.append(round(mean_absolute_error(y_Train, train_preds), 0))
    Test_mae.append(round(mean_absolute_error(y_Test, val_preds), 0))
    oob_scores.append(rf.oob_score_)

In [92]:
#Code added by Keshav[10-03-2021] - Printing all the values for R-Squared and MAE for test and train and out-of-bag score
print("Train r2 scores: \n", Train_r2)
print("")
print("Test r2 scores: \n", Test_r2)
print("")
print("Train MAE scores: \n", Train_mae)
print("")
print("Test MAE scores: \n", Test_mae)
print("")
print("Out-of-bag scores: \n", oob_scores)

Train r2 scores: 
 [0.95, 0.91, 0.84, 0.9, 0.94, 0.95, 0.95, 0.95, 0.87, 0.86]

Test r2 scores: 
 [0.0, 0.23, 0.61, 0.81, 0.0, 0.0, 0.0, 0.65, 0.76, 0.74]

Train MAE scores: 
 [236.0, 335.0, 369.0, 327.0, 233.0, 235.0, 232.0, 291.0, 338.0, 339.0]

Test MAE scores: 
 [1581.0, 506.0, 458.0, 442.0, 1611.0, 1585.0, 1578.0, 487.0, 433.0, 470.0]

Out-of-bag scores: 
 [0.7437992930069671, -0.0001029987046081704, -0.0017979206012916915, 0.0032770430590084443, 0.7017705288471077, 0.7580066555398066, 0.7584123361828632, 0.003442637051243702, 0.0035047349287193885, 0.0029123540422850702]


In [94]:
#Code added by Keshav[10-03-2021]  - Calculating the mean of R-Squared, MAE and out-of-bag score for uncleaned data
print("Mean train r2: ", np.mean(Train_r2))
print("Mean Test r2: ", np.mean(Test_r2))
print("Mean train MAE: ", np.mean(Train_mae))
print("Mean Test MAE: ", np.mean(Test_mae))
print("Mean oob score: ", np.mean(oob_scores))

Mean train r2:  0.9119999999999999
Mean Test r2:  0.38
Mean train MAE:  293.5
Mean Test MAE:  915.1
Mean oob score:  0.2973224663352101


#### Explanation of the code added in the Part 1 [Explaination by Keshav]

In the part 1 - Create and evaluate an initial model we are creating a random forest model and evaluating it on the noisy dataset and checking the MAE and R-Squared score for the training and testing data. to achive this first of all we have a quick look into our dataset using the code "house_rent.info()" and letter on only aquiring the numeric feature or column fromm the dataset for a numeric dataframe because as we are going to use a random forest regressor which uses numeric data to train and test the model. below mentioned steps
are covered in the part 1
1. Loading the pandas package
2. importing the data from the csv file and have a look to the column using the transpose function.
3. Checking the column name, datatype and not-null values using the info function 
4. Extracting only numeric features from the dataset and checking the null values using the ".isnull.any()" function
5. Dividing our dataset into feature vector and Target Vector which is price in this case
6. Importing all the necessary packages for spliting the dataset and for random forest regressor and MAE and R-Squared error
7. Spliting out data into Training and testing[validation] dataset, Creating Random forest regressor and fitting our training dataset
8. Checking the performance of our model using the testing dataset and predicting different errors

### Part 2 - Denoise the data

This section should only include the code necessary to **denoise** the data, NOT the code necessary to identify inconsistencies, problems, errors, etc. in the data. 

#### Code (25 marks)

In [95]:

#Code added by Sankeerth[10-03-2021] - Cleaning the data by removing all unnecessary columns.
Column
_clean = house_rent[['bathrooms', 'bedrooms', 'longitude', 'latitude', 'price']]

#Code added by Sankeerth[10-03-2021] - Cleaning the location data(Using New York Citys longitude range as between 40.65 and 40.95 and the longitude range being between -73.65 and -74.10)
Location_clean = Column_clean[(Column_clean['latitude']<40.95) &
                              (Column_clean['latitude']>40.65) &
                              (Column_clean['longitude']<-73.65) &
                              (Column_clean['longitude']>-74.10)]

#Code added by Sankeerth[10-03-2021] - Cleaning the bedroom data(Considering normal homes in New York City to have 0-5 berooms)
Bedroom_clean = Location_clean[(Location_clean['bedrooms'] <= 5)]

#Code added by Sankeerth[10-03-2021] - Cleaning the bathroom data(Considering normal homes in New York City to have 0-6 bathrooms)
Bathroom_clean = Bedroom_clean[(Bedroom_clean['bathrooms'] <= 6)]

#Code added by Sankeerth[10-03-2021] - Cleaning the Price data(Considering normal homes in New York City to have a price between 1500$ and 12500$)
Clean_Rent_Data = Bathroom_clean[(Bathroom_clean['price'] >= 1500)&
                                 (Bathroom_clean['price'] <= 12500)]

#Code added by Sankeerth[10-03-2021] - Viewing the new dataframe
Clean_Rent_Data.head(10)

#Code added by Sankeerth[10-03-2021] - Viewing the discription of the new dataframe
Clean_Rent_Data.describe()

bathrooms      bedrooms     longitude      latitude         price
count  19261.000000  19261.000000  19261.000000  19261.000000  19261.000000
mean       1.190489      1.523908    -73.972983     40.752129   3534.683609
std        0.448842      1.088652      0.028220      0.035603   1548.269538
min        0.000000      0.000000    -74.048200     40.650300   1500.000000
25%        1.000000      1.000000    -73.991800     40.729600   2500.000000
50%        1.000000      1.000000    -73.978100     40.752100   3150.000000
75%        1.000000      2.000000    -73.955400     40.773800   4065.000000
max        4.000000      5.000000    -73.700100     40.909400  12500.000000

In [96]:
house_rent_num.shape

(20000, 5)

In [97]:
Clean_Rent_Data.shape

(19261, 5)

### Part 3 - Create and evaluate a final model

#### Code (15 marks)

In [98]:
#Code added by Gurpinder Kaur[11-03-2021] - Creating feature vector and Target vector from cleaned data 
X_Clean = Clean_Rent_Data.drop('price', axis=1)
y_Clean = Clean_Rent_Data['price']

In [99]:
#Code added by Gurpinder Kaur[11-03-2021] - Spliting our cleaned datasets into training dataset and testing[validation] datasets
X_Train_Clean, X_Test_Clean, y_Train_Clean, y_Test_Clean = train_test_split(X_Clean, y_Clean, test_size=0.2,random_state =1500)  
#Code added by Gurpinder Kaur[11-03-2021] - Creating a random forest regressor for cleaned data
rf_Clean = RandomForestRegressor(n_estimators=100, n_jobs=-1, oob_score=True, random_state=1500)
#Code added by Gurpinder Kaur[11-03-2021] - Fitting our model using training dataset 
rf_Clean.fit(X_Train_Clean, y_Train_Clean) 
rf_Clean.score(X_Train_Clean, y_Train_Clean)

0.9509686727319452

In [100]:
#Code added by Gurpinder Kaur[11-03-2021] - calculating the out-of-bag score for cleaned data
rf_Clean.oob_score_

0.8086998507797551

In [101]:
#Code added by Gurpinder Kaur[11-03-2021] - Calculating the model score for cleaned data
rf_Clean.score(X_Test_Clean, y_Test_Clean)

0.7898505688702626

In [102]:
#Code added by Gurpinder Kaur[11-03-2021] - Calculating the average price for the training cleaed data and R-Squared score
avg_preds = [np.mean(y_Train_Clean)] * len(y_Test_Clean)

print(np.mean(y_Train_Clean))

r2_score(y_Test_Clean, avg_preds)

3528.254802699896


-0.0004289195519942002

In [103]:
#Code added by Gurpinder Kaur[11-03-2021] - Creating Empty lists for storing the values for testing and training fields respectivelly for cleaned data
Clean_Train_r2 = []
Clean_Train_mae = []
Clean_Test_r2 = []
Clean_Test_mae = []
Clean_oob_scores = []
#Code added by Gurpinder Kaur[11-03-2021] - Creating a for loop to calculate the values of R-Squared, MAE for test and train and Out-of-bag score for 10 runs for cleaned data
for i in range(10):
        X_Train_Clean, X_Test_Clean, y_Train_Clean, y_Test_Clean = train_test_split(X_Clean, y_Clean, test_size=0.2, random_state=1500)  
        rf_Clean = RandomForestRegressor(n_estimators=100, n_jobs=-1, oob_score=True, random_state = 1500)
        rf_Clean.fit(X_Train_Clean, y_Train_Clean) 
        Clean_train_preds = rf_Clean.predict(X_Train_Clean)
        Clean_Test_preds = rf_Clean.predict(X_Test_Clean)
        Clean_Train_r2.append(round(r2_score(y_Train_Clean, Clean_train_preds), 2))
        Clean_Test_r2.append(round(r2_score(y_Test_Clean, Clean_Test_preds), 2))
        Clean_Train_mae.append(round(mean_absolute_error(y_Train_Clean, Clean_train_preds), 0))
        Clean_Test_mae.append(round(mean_absolute_error(y_Test_Clean, Clean_Test_preds), 0))
        Clean_oob_scores.append(rf_Clean.oob_score_)

In [104]:
#Code added by Gurpinder Kaur[11-03-2021] - Printing all the values for R-Squared and MAE for test and train and out-of-bag score
print("Train r2 scores: \n", Clean_Train_r2)
print("")
print("Test r2 scores: \n", Clean_Test_r2)
print("")
print("Train MAE scores: \n", Clean_Train_mae)
print("")
print("Test MAE scores: \n", Clean_Test_mae)
print("")
print("Out-of-bag scores: \n", Clean_oob_scores)

Train r2 scores: 
 [0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95]

Test r2 scores: 
 [0.79, 0.79, 0.79, 0.79, 0.79, 0.79, 0.79, 0.79, 0.79, 0.79]

Train MAE scores: 
 [199.0, 199.0, 199.0, 199.0, 199.0, 199.0, 199.0, 199.0, 199.0, 199.0]

Test MAE scores: 
 [385.0, 385.0, 385.0, 385.0, 385.0, 385.0, 385.0, 385.0, 385.0, 385.0]

Out-of-bag scores: 
 [0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551, 0.8086998507797551]


In [105]:
print("Mean train r2: ", np.mean(Clean_Train_r2))
print("Mean Test r2: ", np.mean(Clean_Test_r2))
print("Mean train MAE: ", np.mean(Clean_Train_mae))
print("Mean Test MAE: ", np.mean(Clean_Test_mae))
print("Mean oob score: ", np.mean(Clean_oob_scores))

Mean train r2:  0.9499999999999998
Mean Test r2:  0.79
Mean train MAE:  199.0
Mean Test MAE:  385.0
Mean oob score:  0.8086998507797551


### Explaination of the Code for Part 3 [By Gurpinder Kaur] ###

The code in the part 3 is almost same as in the part 1 of creating and training a random forest model. The diffirence is in the part 1 we have used the noisy and uncleaned data and in part 3 we have used the cleaned data to see the difference in the results. the "mean out of bag score" using the uncleaned data was 0.297 which shows that the performance of model is no better than just calculating the average price but when we used the cleaned data the "mean out of bag" score is increaded to 0.808 closer to 1 which shows that this model predicts the house prices better than the model in part 1

### Part 4 - Document the problems (35 marks)

In this part, please use the table below to document your understanding of all the data issues you discovered. Note that **no code** should be included, as that should be covered in **Part 2**. Also, note that even if one line of code fixed a few problems, you should list each problem separately in the table below, so be sure you have investigated the data properly. For example, if the list `[-6, 5, 0, 50]` represents heights of adults, the -6, 0, and 50 would represent three data issues to be included in the table below, even though one line of code may be able to address all of them. 

| Data issue discovered | Why is this a problem? | How did you fix it? | Why is this fix appropriate? |
| :- | :- | :- | :- | 
|  example problem 1  | example explanation    | example fix  | example explanation about why this fix is appropriate   |
|  example problem 2  | example explanation    | example fix  | example explanation about why this fix is appropriate   |




















In [111]:
#Code added on 10-03-2021 - Creating a list for the data issues.
Data_Issue = ('Unnamed: 0 column', 
              'number of bathrooms',
              'number of bedrooms',
              'building_id column',
              'created column',
              'description column',
              'display_address column',
              'features column',
              'upper limit of the latitude',
              'lower limit of the latitude',
              'upper limit of the longitude',
              'lower limit of the longitude',
              'manager_id column',
              'photos column',
              'lower limit of the price',
              'upper limit of the price',
              'street_address column',
              'interest_level column',
              'mgr_apt_count column')  
#Code added on 10-03-2021 - Creating a list for why the data issue is a problem.
Problem = ('This column is not pertinent to the business Question',
           'There are outliers in this data column',
           'There are outliers in this data column',
           'This column is not pertinent to the business Question and is non-numerical data',
           'This is the date of creation in the database and does not help us answer the business question',
           'This is a non-numerical data',
           'This is a non-numerical data',
           'This is a list and non-numerical data',
           'There seem to be latitudes from ouside New York City',
           'There seem to be latitudes from ouside New York Cit',
           'There seem to be longitudes from ouside New York Cit',
           'There seem to be longitudes from ouside New York Cit',
           'This column is not pertinent to the business Question and is non-numerical data',
           'This is a non-numerical data',
           'Existance of outliers',
           'Existance of outliers',
           'This is a non-numerical data',
           'This is a non-numerical data',
           'This column is not pertinent to the business Question')
#Code added on 10-03-2021 - Creating a list for how the problem was solved.
Fix = ('Removing Column',
       'Settring an upper limit the the number of bedrooms in a house',
       'Settring an upper limit the the number of bathrooms in a house',
       'Removing Column',
       'Removing Column',
       'Removing Column',
       'Removing Column',
       'Removing Column',
       'Setting an upper limit for the latitudes',
       'Setting an lower limit for the latitudes',
       'Setting an upper limit for the longitudes',
       'Setting an lower limit for the longitudes',
       'Removing Column',
       'Removing Column',
       'Settring an lower limit for the rent price in New York city',
       'Settring an upper limit for the rent price in New York city',
       'Removing Column',
       'Removing Column',
       'Removing Column')
#Code added on 10-03-2021 - Creating a list for why the solution was appropriate.
Why_Fix = ('As Unnamed: 0 column just contains the serial numbers for the rows it did not help dermine the rent price for the building so the column was removed.',
           'As buildings with more than 5 bedrooms are rare in New York city and are outliers to the data we decided to remove all buildings with more than 5 bedrooms.',
           'As buildings with more than 6 bathrooms are rare in New York city and are outliers to the data we decided to remove all buildings with more than 6 bathrooms.',
           'As building_id is a refernece for the databases creators and is non-numerical. This feature does not help determine the rent price for the building so the column was removed.',
           'This is the date of creation in the database. As this does not effect the rent price for the building so the column was removed.',
           'As the data is non-numerical(i.e. discription of the building) the column can not be read by the model. Thus, the column was removed',
           'As the data is non-numerical the column can not be read by the model.Furthermore as Longitude and Latitude is used for to identify the location this colums was not needed and was removed',
           'As the data is a list of non_numerical data the column can not be read by the model. Thus, the column was removed',
           'The upper limit of the latitude is set at 40.95(aprox latitude of the northern tip of New York City) so that we dont get any data from building to the north of New York City',
           'The lower limit of the latitude is set at 40.65(aprox latitude of the southern tip of New York City) so that we dont get any data from building to the south of New York City',
           'The upper limit of the latitude is set at -73.65(aprox longitude of the eastern tip of New York City) so that we dont get any data from building to the east of New York City',
           'The lower limit of the latitude is set at -74.10(aprox longitude of the western tip of New York City) so that we dont get any data from building to the west of New York City',
           'As manager_id is a refernece(i.e. for the manager of the building) for the databases creators and is non-numerical. This feature does not help determine the rent price for the building so it was removed',
           'As the data is non-numerical(i.e image link) the column can not be read by the model. Thus, the column was removed',
           'As any building with a rent less than 1000$ in New York may not be a house that most people would want to live in. So we decided to remove all houses witha price less tham 1000$',
           'As any building with a price over 12500$ in New York city only caters to a veru niche segment and can be considered as an outlier to our data. So we decided to remove all houses witha price higher tham 12500$',
           'As the data is non-numerical the column can not be read by the model.Furthermore as Longitude and Latitude is used for to identify the location this colums was not needed and was removed',
           'As the data is non-numerical(i.e interest levels in the buildings) the column can not be read by the model. Thus, the column was removed',
           'As this colums talks about the management rather than the building so it does not help ermine the rent price for the building. Thus, this column was removed')
#Code added on 10-03-2021 - Creating a dataframe.
Part4 = pd.DataFrame({'Data issue discovered' : Data_Issue,
                      'Why is this a problem?' : Problem,
                      'How did you fix it?' : Fix,
                      'Why is this fix appropriate?' : Why_Fix},
                      columns = ['Data issue discovered','Why is this a problem?','How did you fix it?','Why is this fix appropriate?'])
#Code added on 10-03-2021 - Printing out the dataframe.
Part4



Data issue discovered  \
0              Unnamed: 0 column   
1            number of bathrooms   
2             number of bedrooms   
3             building_id column   
4                 created column   
5             description column   
6         display_address column   
7                features column   
8    upper limit of the latitude   
9    lower limit of the latitude   
10  upper limit of the longitude   
11  lower limit of the longitude   
12             manager_id column   
13                 photos column   
14      lower limit of the price   
15      upper limit of the price   
16         street_address column   
17         interest_level column   
18          mgr_apt_count column   

                               Why is this a problem?  \
0   This column is not pertinent to the business Q...   
1              There are outliers in this data column   
2              There are outliers in this data column   
3   This column is not pertinent to the business Q...   
4   This is the date of creation in the database a...   
5                        This is a non-numerical data   
6                        This is a non-numerical data   
7               This is a list and non-numerical data   
8   There seem to be latitudes from ouside New Yor...   
9   There seem to be latitudes from ouside New Yor...   
10  There seem to be longitudes from ouside New Yo...   
11  There seem to be longitudes from ouside New Yo...   
12  This column is not pertinent to the business Q...   
13                       This is a non-numerical data   
14                              Existance of outliers   
15                              Existance of outliers   
16                       This is a non-numerical data   
17                       This is a non-numerical data   
18  This column is not pertinent to the business Q...   

                                  How did you fix it?  \
0                                     Removing Column   
1   Settring an upper limit the the number of bedr...   
2   Settring an upper limit the the number of bath...   
3                                     Removing Column   
4                                     Removing Column   
5                                     Removing Column   
6                                     Removing Column   
7                                     Removing Column   
8            Setting an upper limit for the latitudes   
9            Setting an lower limit for the latitudes   
10          Setting an upper limit for the longitudes   
11          Setting an lower limit for the longitudes   
12                                    Removing Column   
13                                    Removing Column   
14  Settring an lower limit for the rent price in ...   
15  Settring an upper limit for the rent price in ...   
16                                    Removing Column   
17                                    Removing Column   
18                                    Removing Column   

                         Why is this fix appropriate?  
0   As Unnamed: 0 column just contains the serial ...  
1   As buildings with more than 5 bedrooms are rar...  
2   As buildings with more than 6 bathrooms are ra...  
3   As building_id is a refernece for the database...  
4   This is the date of creation in the database. ...  
5   As the data is non-numerical(i.e. discription ...  
6   As the data is non-numerical the column can no...  
7   As the data is a list of non_numerical data th...  
8   The upper limit of the latitude is set at 40.9...  
9   The lower limit of the latitude is set at 40.6...  
10  The upper limit of the latitude is set at -73....  
11  The lower limit of the latitude is set at -74....  
12  As manager_id is a refernece(i.e. for the mana...  
13  As the data is non-numerical(i.e image link) t...  
14  As any building with a rent less than 1000$ in...  
15  As any building with a price over 12500$ in Ne...  
16  As the data is non-numerical the column can no...  
